src/hooks/useSearchParamState.ts

```diff
@@ -105,13 +110,56 @@ const getAllCurrentParams = (searchParams: URLSearchParams) => {
   return allUrlParams;
 };
 
-const useSearchParamsState = (paramsDefinition: SearchParamsStateType) => {
+/**
+ * Works like useState but stores the state in URL search params. Adapted
+ * from https://github.com/jschwindt/react-use-search-params-state/tree/main
+ *
+ * @param paramsDefinition maps search param names to their definitions,
+ * indicating their type and default value
+ * @param initial initial values to set in the search params
+ */
+const useSearchParamsState = ({
+  paramsDefinition,
+  initial,
+}: {
+  paramsDefinition: SearchParamsStateType;
+  initial?: URLSearchParamsInit;
+}) => {
+  // `initial` is NOT passed to `useSearchParams` here, since react-router-dom's
+  // useSearchParams's `defaultInit` prop auto-populates the internal state, but not
+  // the search params that appear in the url. See the useEffect below
   const [searchParams, setSearchParams] = useSearchParams();
 
-  const values = useMemo(
-    () => getValues(paramsDefinition, searchParams),
-    [paramsDefinition, searchParams]
-  );
+  const [mounted, setMounted] = useState(false);
+
+  // Use an effect to set the `initial` params in the url bar
+  useEffect(() => {
+    setMounted(true);
+
+    if (!initial) return;
+
+    const toUpdate: Record<string, any> = Object.fromEntries(
+      // Find any key/value pairs that are in initial, but not in searchParams
+      Object.entries(initial).filter(([key]) => !searchParams.has(key))
+    );
+    if (Object.keys(toUpdate).length) {
+      // Spread currentParams to avoid overwriting search params set by other components
+      const currentParams = getAllCurrentParams(searchParams);
```

Comment:
The longer I stare at this, the more I think there might be latent bugs with providing `initial` values for param types that are not strings. (Latent because we aren't currently doing this.) I'll check again tomorrow to see if I can get a bit more confident about the behavior. But I also think it might be nice to just merge this into 154 regardless, so that the existing behavior can be QA'd along with the other optional column work.

Reply:
Ok, after taking another look this morning, this and my other outstanding comment above are related: https://github.com/greenriver/hmis-frontend/pull/1076#discussion_r1982640872

- `initial` is typed as a `URLSearchParamsInit`, which enforces that initial values are strings
- If we were to instead remove the `initial` arg to the hook and add an `initial` prop alongside `default` as part of the `SearchParamStateType`, then we _would_ need to convert any non-strings into string values. (Maybe by adapting/reusing the code [here](https://github.com/greenriver/hmis-frontend/pull/1076/files#diff-730f4ec2e5f9d2f89a0a784e17cc19f00aeeb106526cb26c74af2275db1f4bc7R167-R187))
- I'm not sure that is desirable or more clear, plus am running out of time on this ticket, so I'd propose to merge this and leave any future improvements un-ticketed until the need arises.

Reply:
I guess I'm a little confused about the decision to type `initial` as a `URLSearchParamsInit`, as opposed to typing it as the same type as `value`. Can you explain a little bit?

Again, thinking that we want this hook to be analogous to `const [value, setValue] = useState(initial)`, then we would want the `initial` to have the same type as `value`.

For example:

```tsx
// Type for internal state representation of search params
type ValueType = Record<string, any>; // Future improvement: constrain value type
// Arg type for this hook
type Args = {
  paramsDefinition: SearchParamsStateType; // shape of params/state
  initial?: ValueType; // initial value
};
// Returns state and setState function
type ReturnType = [ValueType, (value: ValueType) => void];

const useSearchParamsState = ({
  paramsDefinition,
  initial,
}: Args): ReturnType => {
...
```

If the `initial` arg represented the initial values in the  'state' representation -- rather then the initial values in a URLSearchParamsInit representation -- then maybe we can also get rid of the transformation `getValues(paramsDefinition, createSearchParams(initial))` which converts `initial` into SearchParams and then back into "values".  Instead we could use `initial` directly, I think, since it's already the right shape. Does that make sense?


> If we were to instead remove the initial arg to the hook and add an initial prop alongside default as part of the SearchParamStateType

I'm not totally following this. API-wise, this doesn't seem desirable to me (again trying to mimic somewhat the useState API)

Reply:
Hm thank you for the feedback Gig. I added your suggestions https://github.com/greenriver/hmis-frontend/pull/1076/commits/666b8cf25615681449f6dd1fb61950c18898dc79, let me know if it looks ok to you.

> the decision to type initial as a URLSearchParamsInit, as opposed to typing it as the same type as value. Can you explain a little bit?

no reason, I'd just typed it that way in original implementation when I thought it would be passed in to the `useSearchParams` hook as the `initial` arg, but that was removed.

> I'm not totally following this. API-wise, this doesn't seem desirable to me (again trying to mimic somewhat the useState API)

just to explain my thinking a bit more, one thing I thought was messy about the API where usage looks like this:
```
  const [paramValues, setParamValues] = useSearchParamsState({
    paramsDefinition: {
      optionalColumns: {
        type: 'string',
        multiple: true,
        default: [],
      },
    },
    initial: {
      optionalColumns: initialOptionalColumns,
    },
  });
```
is that we separately declare `optionalColumns` on both objects passed to the hook. And I thought it would be better for usage to look more like this:
```
  const [paramValues, setParamValues] = useSearchParamsState({
    paramsDefinition: {
      optionalColumns: {
        type: 'string',
        multiple: true,
        default: [],
        initial: initialOptionalColumns
      },
    },
  });
```

Part of my slowness here was that by "mimicking the useState API" I got stuck on thinking you meant changing the hook to deal with one piece of state at a time, something more like
```
const useSearchParamsState<T> = (initial: T) => {
  ...
} 
```
and the usage would be more like
```
// in bulk services
const [coc, setCoc] = useSearchParamsState<string>();
const [serviceDate, setServiceDate] = useSearchParamsState<date>();
...

// or in optional columns land,
const [optionalColumns, setOptionalColumns] = useSearchParamsState<string[]>(initialOptionalColumns);
```
but eventually I figured out that is a bigger refactor than you are proposing here.

Reply:
Thanks @martha, sorry for my lack of clarity there. The type changes look good to me, can we remove the `createSearchParams` transformation or is that still necessary? Seems like another area that could be potentially bug-prone with other non-string types.

Reply:
No no worries at all, I'm sure it's me 😅!

Removing `createSearchParams` is a bit tricky since `getValues` accepts a `URLSearchParams` object. I think we could refactor so that it accepts either `UrlSearchParams` or a `Record<string, any>` but I'm not totally sure I see the benefit, since inside of all the `paramsToType` methods we would then need to check the type of `searchParams` every time it's used to figure out how to get what's inside it. That seems like it would add some cruft/lack of code clarity that I am a bit hesitant about.

Reply:
Right but the `initial` now has the same type as "value" (ValueType), so I'm wondering why we need to transform it to search params and then back to a value object? The code is:

```ts
    if (initial && !mounted) {
      // If the above useEffect hasn't run yet, return the initial values right away
      // to avoid flickering
      return getValues(paramsDefinition, createSearchParams(initial));
    }
```

and I'm wondering whether it can be
```suggestion
    if (initial && !mounted) {
      // If the above useEffect hasn't run yet, return the initial values right away
      // to avoid flickering
      + return initial;
    }
```